internal/controller/factory/pdb.go

```diff
@@ -21,49 +21,49 @@ import (
 	"fmt"
 
 	etcdaenixiov1alpha1 "github.com/aenix-io/etcd-operator/api/v1alpha1"
+	"github.com/aenix-io/etcd-operator/internal/log"
 	v1 "k8s.io/api/policy/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/intstr"
 	"k8s.io/utils/ptr"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
-	"sigs.k8s.io/controller-runtime/pkg/log"
 )
 
 func CreateOrUpdatePdb(
 	ctx context.Context,
 	cluster *etcdaenixiov1alpha1.EtcdCluster,
 	rclient client.Client,
 ) error {
-	if cluster.Spec.PodDisruptionBudgetTemplate == nil {
-		return deleteOwnedResource(ctx, rclient, &v1.PodDisruptionBudget{
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: cluster.Namespace,
-				Name:      cluster.Name,
-			}})
-	}
-
-	logger := log.FromContext(ctx)
 	pdb := &v1.PodDisruptionBudget{
 		ObjectMeta: metav1.ObjectMeta{
 			Namespace: cluster.Namespace,
 			Name:      cluster.Name,
 		},
-		Spec: v1.PodDisruptionBudgetSpec{
-			MinAvailable:   cluster.Spec.PodDisruptionBudgetTemplate.Spec.MinAvailable,
-			MaxUnavailable: cluster.Spec.PodDisruptionBudgetTemplate.Spec.MaxUnavailable,
-			Selector: &metav1.LabelSelector{
-				MatchLabels: NewLabelsBuilder().WithName().WithInstance(cluster.Name).WithManagedBy(),
-			},
-			UnhealthyPodEvictionPolicy: ptr.To(v1.IfHealthyBudget),
+	}
+	ctx, err := contextWithGVK(ctx, pdb, rclient.Scheme())
+	if err != nil {
+		return err
+	}
+
+	if cluster.Spec.PodDisruptionBudgetTemplate == nil {
+		return deleteOwnedResource(ctx, rclient, pdb)
+	}
+
+	pdb.Spec = v1.PodDisruptionBudgetSpec{
```

Comment:
As I understood you restructured this piece of code a bit to avoid duplication of 
```
ObjectMeta: metav1.ObjectMeta{
				Namespace: cluster.Namespace,
				Name:      cluster.Name,
			}
```

Is it possible to achieve the same without separation of pdb object to ObjectMeta and pdb.Spec?

The problem is that when such object structures are separated in the code, in the end it is sometimes more complicated to search them in different pieces of code. In this case they are close to each other, but in the future they can go away from each other and be placed in different pieces of the go file. Readability suffers in my opinion.

From my perspective always easier to have resulted object describe in one place, than combine objects like here: https://github.com/aenix-io/etcd-operator/blob/010228b82e26a605f78449a5fbfd7350f0fd6475/internal/controller/factory/statefulset.go#L358

Reply:
@Kirill-Garbar totally agree with you. Unfortunately, I didn't find a way how to do it without making this piece of code even more ugly. The only option which might be suitable from my perspective is to manually set context values in case of pdb deletion:
```go
	if cluster.Spec.PodDisruptionBudgetTemplate == nil {
		ctx = log.WithValues(ctx, "group", "v1", "kind", "PodDisruptionBudget", "name", cluster.Name)
		return deleteOwnedResource(ctx, rclient, &v1.PodDisruptionBudget{
			ObjectMeta: metav1.ObjectMeta{
				Namespace: cluster.Namespace,
				Name:      cluster.Name,
			},
		})
	}
```
Not the best solution, but still. WDYT?

Reply:
I think it's fine. I do not have better idea.